core/src/main/scala/pureconfig/ConfigCursor.scala

```diff
@@ -201,6 +249,69 @@ object ConfigCursor {
    * @return a `ConfigCursor` for `value` at the path given by `pathElems`.
    */
   def apply(value: ConfigValue, pathElems: List[String]): ConfigCursor = SimpleConfigCursor(value, pathElems)
+
+  /**
+   * Handle automatic type conversions of `ConfigValue`s the way the
+   * [[https://github.com/lightbend/config/blob/master/HOCON.md#automatic-type-conversions HOCON specification]]
+   * describes. This code mimics the behavior of the package-private `com.typesafe.config.impl.DefaultTransformer` class
+   * in Typesafe Config.
+   */
+  private[pureconfig] def transform(configValue: ConfigValue, requested: ConfigValueType): Either[WrongType, ConfigValue] = {
+    (configValue.valueType(), requested) match {
+      case (valueType, requestedType) if valueType == requestedType =>
+        Right(configValue)
+
+      case (ConfigValueType.STRING, requestedType) =>
+        val s = configValue.unwrapped.asInstanceOf[String]
+
+        requestedType match {
+          case ConfigValueType.NUMBER =>
+            lazy val tryLong = Try(s.toLong).map(ConfigValueFactory.fromAnyRef)
+            lazy val tryDouble = Try(s.toDouble).map(ConfigValueFactory.fromAnyRef)
+            // Try#toEither is only available in Scala 2.12+.
+            (tryLong orElse tryDouble) match {
+              case Success(value) => Right(value)
+              case Failure(_) => Left(WrongType(configValue.valueType, Set(NUMBER)))
+            }
+
+          case ConfigValueType.NULL if s == "null" =>
+            Right(ConfigValueFactory.fromAnyRef(null))
+
+          case ConfigValueType.BOOLEAN if s == "true" || s == "yes" || s == "on" =>
+            Right(ConfigValueFactory.fromAnyRef(true))
+
+          case ConfigValueType.BOOLEAN if s == "false" || s == "no" || s == "off" =>
+            Right(ConfigValueFactory.fromAnyRef(false))
+
+          case other =>
+            Left(WrongType(configValue.valueType(), Set(other)))
+        }
+
+      case (ConfigValueType.NUMBER | ConfigValueType.BOOLEAN, ConfigValueType.STRING) =>
+        Right(ConfigValueFactory.fromAnyRef(configValue.unwrapped.toString))
+
+      case (ConfigValueType.OBJECT, ConfigValueType.LIST) =>
+        val obj = configValue.asInstanceOf[ConfigObject].asScala
+        val ll = obj.foldLeft[Either[List[String], List[(Int, ConfigValue)]]](Right(Nil)) {
+          case (acc, (str, v)) =>
+            (acc, Try(str.toInt)) match {
+              case (Right(a), Success(b)) => Right((b -> v) :: a)
+              case (Left(keys), Failure(_)) => Left(str :: keys)
```

Comment:
If I understand this well, `acc` never becomes a `Left` and even if it did, are not doing anything with the invalid keys. Can we simplify this?

(I didn't know HOCON implementations should ignore non-integer keys in this case. What a strange behavior...)

Reply:
I ended up naively changing the old code (in which we used the keys) to fully support the HOCON specification. This can indeed be simplified. I did that in d6f0b84.